worker/secretrotate/secretrotate.go

```diff
@@ -0,0 +1,222 @@
+	changes, err := w.config.SecretManagerFacade.WatchSecretsRotationChanges(w.config.SecretOwner.String())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(changes); err != nil {
+		return errors.Trace(err)
+	}
+	for {
+		var timeout <-chan time.Time
+		if w.timer != nil {
+			timeout = w.timer.Chan()
+		}
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case ch, ok := <-changes.Changes():
+			if !ok {
+				return errors.New("secret rotation change channel closed")
+			}
+			w.handleSecretRotateChanges(ch)
+		case now := <-timeout:
+			w.rotate(now)
+		}
+	}
+}
+
+func (w *Worker) rotate(now time.Time) {
+	w.config.Logger.Tracef("processing secret rotation for %q at %v", w.config.SecretOwner, now)
+
+	var toRotate []string
+	for id, info := range w.secrets {
+		if !info.rotateTime.After(now) {
+			toRotate = append(toRotate, info.URL.ShortString())
+			// Once secret has been queued for rotation, delete it here since
+			// it will re-appear via the watcher after the rotation is actually
+			// performed and the last rotated time is updated.
+			delete(w.secrets, id)
+		}
+	}
+
+	if len(toRotate) > 0 {
+		w.config.RotateSecrets <- toRotate
+	}
+	w.nextTimeout = time.Time{}
+	w.computeNextRotateTime()
+}
+
+func (w *Worker) handleSecretRotateChanges(changes []watcher.SecretRotationChange) {
+	w.config.Logger.Tracef("got rotate secret changes: %# v", pretty.Formatter(changes))
+	if len(changes) == 0 {
+		return
+	}
+
+	for _, ch := range changes {
+		// Rotate interval of 0 means the rotation has been deleted.
+		if ch.RotateInterval == 0 {
+			delete(w.secrets, ch.ID)
+			continue
+		}
+		w.secrets[ch.ID] = secretRotateInfo{
+			URL:        ch.URL,
+			rotateTime: ch.LastRotateTime.Add(ch.RotateInterval),
+		}
+	}
+	w.computeNextRotateTime()
+}
+
+func (w *Worker) computeNextRotateTime() {
+	w.config.Logger.Tracef("computing next rotated time for secrets %# v", pretty.Formatter(w.secrets))
+
+	if len(w.secrets) == 0 {
+		w.timer = nil
+		return
+	}
+
+	now := w.config.Clock.Now()
+	var nextTick time.Time
+	for _, info := range w.secrets {
+		if !nextTick.IsZero() && info.rotateTime.After(nextTick) {
+			continue
+		}
+		nextTick = info.rotateTime
+	}
+
+	// Account for the worker not running when a secret
+	// should have been rotated.
+	if !nextTick.After(now) {
+		nextTick = now
+	}
+
+	if !w.nextTimeout.IsZero() && nextTick.After(w.nextTimeout) {
+		w.config.Logger.Tracef("retaining rotate time for next secret for %q: will rotate at %v", w.config.SecretOwner, w.nextTimeout)
+		return
+	}
+
+	w.nextTimeout = nextTick
+	// A one minute granularity is acceptable for secret rotation.
+	nextDuration := w.nextTimeout.Sub(now).Round(time.Minute)
+	w.config.Logger.Debugf("next secret for %q will rotate in %v at %v", w.config.SecretOwner, nextDuration, w.nextTimeout)
+
+	if w.timer == nil {
+		w.timer = w.config.Clock.NewTimer(nextDuration)
+	} else {
+		if !w.timer.Stop() {
+			select {
```

Comment:
I don't get this, a comment here would be nice.

Reply:
Added a comment